turtle_tf2_cpp/src/turtle_tf2_listener.cpp

```diff
@@ -0,0 +1,144 @@
+#include <tf2_ros/transform_listener.h>
+#include <tf2_ros/buffer.h>
+# include <turtlesim/srv/spawn.hpp>
+
+#include <chrono>
+#include <memory>
+#include <string>
+
+using std::placeholders::_1;
+using namespace std::chrono_literals;
+
+class FrameListener : public rclcpp::Node
+{
+public:
+  FrameListener()
+  : Node("turtle_tf2_frame_listener")
+  {
+    // Declare and acquire `target_frame` parameter
+    this->declare_parameter<std::string>("target_frame", "turtle1");
+    this->get_parameter("target_frame", target_frame_);
+
+    tf_buffer_ =
+      std::make_unique<tf2_ros::Buffer>(this->get_clock());
+    transform_listener_ =
+      std::make_shared<tf2_ros::TransformListener>(*tf_buffer_);
+
+    // Create a client to spawn a turtle
+    rclcpp::Client<turtlesim::srv::Spawn>::SharedPtr spawner =
+      this->create_client<turtlesim::srv::Spawn>("spawn");
+
+    // Check if the service is available
+    while (!spawner->wait_for_service(1s)) {
+      if (!rclcpp::ok()) {
+        RCLCPP_ERROR(
+          this->get_logger(),
+          "Interrupted while waiting for the service. Exiting."
+        );
+        continue;
+      }
+      RCLCPP_INFO(
+        this->get_logger(),
+        "Service not available, waiting again..."
+      );
+    }
+
+    // Initialize request with turtle name and coordinates
+    // Note that x, y and theta are defined as floats in turtlesim/srv/Spawn
+    auto request = std::make_shared<turtlesim::srv::Spawn::Request>();
+    request->x = 4.0;
+    request->y = 2.0;
+    request->theta = 0.0;
+    request->name = "turtle2";
+    // Call request
+    auto result = spawner->async_send_request(request);
+
+    // Create turtle2 velocity publisher
+    publisher_ = this->create_publisher<geometry_msgs::msg::Twist>("turtle2/cmd_vel", 1);
+
+    // Call on_timer function every second
+    timer_ = this->create_wall_timer(
+      1s, std::bind(&FrameListener::on_timer, this));
+  }
+
+private:
+  void on_timer()
+  {
+    // Store frame names in variables that will be used to
+    // compute transformations
+    std::string from_frame_rel = target_frame_.c_str();
+    std::string to_frame_rel = "turtle2";
+
+    geometry_msgs::msg::TransformStamped transformStamped;
+
+    // Look up for the transformation between target_frame and turtle2 frames
+    // and send velocity commands for turtle2 to reach target_frame
+    try {
+      transformStamped = tf_buffer_->lookupTransform(
+        to_frame_rel, from_frame_rel,
+        tf2::TimePoint(),
+        500ms);
+    } catch (tf2::LookupException & ex) {
+      RCLCPP_INFO(this->get_logger(), "transform not ready");
+      return;
+    }
+
+    geometry_msgs::msg::Twist msg;
+    msg.angular.z = 1.0 * atan2(
+      transformStamped.transform.translation.y,
+      transformStamped.transform.translation.x);
+
+    msg.linear.x = 0.5 * sqrt(
+      pow(transformStamped.transform.translation.x, 2) +
+      pow(transformStamped.transform.translation.y, 2));
```

Comment:
It'd be nice to explain what you're doing here. I see that you're taking the L2 norm. I'm not sure why it's multiplied by 0.5.

Reply:
Same as the comment above.

Reply:
Resolved in https://github.com/ros/geometry_tutorials/pull/44/commits/6026a2eaf84a51c4ad8f59234f08993a62258b16